### PR TITLE
Replace CreateConnectionCallback with std::function callbacks

### PR DESCRIPTION
**Pull request by pepone**

### CODE REVIEW

**Review by externl — APPROVED** [review]

File: cpp/src/Ice/ConnectionFactory.cpp [hunk 1]
```diff
@@ -227,17 +229,24 @@ IceInternal::OutgoingConnectionFactory::create(const vector<EndpointIPtr>& endpt
         Ice::ConnectionIPtr connection = findConnection(endpoints, compress);
         if(connection)
         {
-            callback->setConnection(connection, compress);
+            response(connection, compress);
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionFactory.cpp:232]: minor: `std::move(connection)`

File: cpp/src/Ice/ConnectionFactory.cpp [hunk 2]
```diff
@@ -227,17 +229,24 @@ IceInternal::OutgoingConnectionFactory::create(const vector<EndpointIPtr>& endpt
...
             return;
         }
     }
     catch (const std::exception&)
     {
-        callback->setException(current_exception());
+        exception(current_exception());
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionFactory.cpp:238]: I would use `catch (...)` here given how we handle it.

File: cpp/src/Ice/ConnectionFactory.cpp [hunk 3]
```diff
@@ -227,17 +229,24 @@ IceInternal::OutgoingConnectionFactory::create(const vector<EndpointIPtr>& endpt
...
         return;
     }
 
-    auto cb = make_shared<ConnectCallback>(_instance, shared_from_this(), endpoints, hasMore, callback, selType);
+    auto cb = make_shared<ConnectCallback>(
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionFactory.cpp:242]: Next PR is removing ConnectCallback?

File: cpp/src/Ice/Reference.cpp [hunk 4]
```diff
@@ -1549,136 +1549,74 @@ IceInternal::RoutableReference::createConnectionAsync(
...
-                if(_reference->getRouterInfo() && _reference->getRouterInfo()->getAdapter())
-                {
-                    connection->setAdapter(_reference->getRouterInfo()->getAdapter());
-                }
-                _response(connection, compress);
-            }
-
-            void setException(std::exception_ptr ex) final
+        auto state = make_shared<CreateConnectionState>();
+        state->createConnectionException =
```
Comment (bernardnormier) [inline cpp/src/Ice/Reference.cpp:1592]: Is this code leak-free?

My concern:
- CreateConnectionState holds onto std::function object (createConnetionException)
- std::function object own a refcount on shared_ptr<CreateConnectionState> (in lambda capture)

File: cpp/src/Ice/ConnectionFactory.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/Ice/ConnectionFactory.cpp:242]: Yes, it will

File: cpp/src/Ice/Reference.cpp (hunk 4 above)
Comment (pepone) [reply cpp/src/Ice/Reference.cpp:1615]: Updated the code to clear `CreateConnectionState`

**Comment by pepone** [comment]
Not clear why IceStorm/repstress is failing with this PR, seems is hanging on shutdown.

File: cpp/src/Ice/Reference.cpp [hunk 5]
```diff
@@ -1572,137 +1572,85 @@ IceInternal::RoutableReference::createConnectionAsync(
...
 
-        //
-        // Get an existing connection or create one if there's no
-        // existing connection to one of the given endpoints.
-        //
-        factory->create(endpoints, false, getEndpointSelection(), make_shared<CB1>(_routerInfo, response, exception));
-        return;
+    if(getCacheConnection() || endpoints.size() == 1)
+    {
+        // Get an existing connection or create one if there's no existing connection to one of the given endpoints.
```
Comment (bernardnormier) [inline cpp/src/Ice/Reference.cpp:1592]: I find this code too complicated, with a lambda that captures a response and itself and then needs to clear itself.

There must a clearer way to write all this.

File: cpp/src/Ice/Reference.cpp [hunk 6]
```diff
@@ -1575,137 +1575,83 @@ IceInternal::RoutableReference::createConnectionAsync(
...
-        // connection for one of the endpoints.
-        //
-
-        vector<EndpointIPtr> endpt;
-        endpt.push_back(endpoints[0]);
-        RoutableReferencePtr self = dynamic_pointer_cast<RoutableReference>(
-            const_cast<RoutableReference*>(this)->shared_from_this());
-        factory->create(endpt, true, getEndpointSelection(), make_shared<CB2>(self, endpoints, response, exception));
-        return;
+        auto state = make_shared<CreateConnectionState>();
```
Comment (bernardnormier) [inline cpp/src/Ice/Reference.cpp:1642]: Give CreateConnetionState a constructor?

**Review by bernardnormier — APPROVED** [review]
Much clearer than the previous version.

**Pull request merged**